### PR TITLE
Add ServerEndpoint support for Azure AD scenario

### LINKED ISSUE #1802: https://{name}.service.signalr.net/api/v1/auth/accessKey not proxied with ServerEndpoint or ClientEndpoint properties
### Describe the bug
We are currently using SignalR in default mode, and have set up a proxy endpoint for SignalR in API Management by following the instructions in this link: https://learn.microsoft.com/en-us/azure/azure-signalr/signalr-howto-work-with-apim. Additionally, we have added an extra endpoint in API Management for ServerEndpoint Websocket connection, which has a custom policy to replace the current Authorization header with API Management managed identity access token. This means that you only need to add the API Management managed identity to the "SignalR App Server" role, and anyone can connect as a server using the API Management proxy. The server authentication is happening in the API Management policy only.

However, the issue we are facing now is that the SDK is calling https://{name}.service.signalr.net/api/v1/auth/accessKey directly, instead of using the ServerEndpoint or ClientEndpoint proxy. As a result, the solution is failing to negotiate client connection requests.

### To Reproduce
Follow the instructions provided in this Microsoft Azure SignalR tutorial: https://learn.microsoft.com/en-us/azure/azure-signalr/signalr-howto-work-with-apim.

Additionally, create a websocket API with a URL path of /server

**AspNet Core Server Code:**
```
services.AddSignalR(configure =>
{
    configure.EnableDetailedErrors = true;
})
.AddAzureSignalR(option =>
{
    option.Endpoints = new ServiceEndpoint[]
    {
         new ServiceEndpoint(new Uri("https://{name}.service.signalr.net"), new DefaultAzureCredential())
         {
              ClientEndpoint = new Uri("..."),
              ServerEndpoint = new Uri("..."),
         }
     };
});

....

app.UseEndpoints(endpoints =>
{
     endpoints.MapHub<MessageHub>("/messageHub");
});
```
**Client Code:**
```
_connectionBuilder = new HubConnectionBuilder();
_connection = _connectionBuilder
   .WithUrl("http://localhost:5000/messageHub")
   .WithAutomaticReconnect()
   .Build();
```

Run the server and client solutions and attempt to establish a connection.

**We will close this issue if:**
- Proxy https://{name}.service.signalr.net/api/v1/auth/accessKey using API Management ServerEndpoint or ClientEndpoint URL.
- /api/v1/auth/accessKey should append to full proxy URL not just base URL
- find any other solution to fix this issue

### Exceptions (if any)
```
info: Microsoft.Azure.SignalR.AadAccessKey[3]
      Succeed in authorizing AccessKey for 'https://{name}.service.signalr.net/'
crit: Microsoft.Azure.SignalR.ServiceRouteHelper[1]
      Client negotiate failed: The given AzureAD identity don't have the permission to generate access token.
info: Microsoft.Azure.SignalR.AadAccessKey[3]
      Succeed in authorizing AccessKey for 'https://{name}.service.signalr.net/'
info: Microsoft.Azure.SignalR.AadAccessKey[3]
      Succeed in authorizing AccessKey for 'https://{name}.service.signalr.net/'
```

### Further technical details
- `<PackageReference Include="Microsoft.Azure.SignalR" Version="1.21.3" />`
- `<FrameworkReference Include="Microsoft.AspNetCore.App" />`
- `<PackageReference Include="Azure.Identity" Version="1.9.0" />`
- `<PackageReference Include="Microsoft.AspNetCore.SignalR.Client" Version="7.0.7" />`

### PR DESCRIPTION
**Pull request by terencefan**
### Summary of the changes (Less than 80 chars)
 - Add ServerEndpoint support for Azure AD scenario 

Fixes #bugnumber (in this specific format)

Fix #1802

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/Endpoints/AadAccessKey.cs [hunk 1]
```diff
@@ -61,14 +61,10 @@ private set
...
-        public AadAccessKey(Uri uri, TokenCredential credential) : base(uri)
+        public AadAccessKey(Uri endpoint, TokenCredential credential, Uri serverEndpoint = null) : base(endpoint)
         {
-            var builder = new UriBuilder(Endpoint)
-            {
-                Path = "/api/v1/auth/accessKey",
-                Port = uri.Port
-            };
-            AuthorizeUrl = builder.Uri.AbsoluteUri;
+            var authorizeUri = (serverEndpoint ?? endpoint).Append("/api/v1/auth/accessKey");
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/AadAccessKey.cs:66]: is Port missing?

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs [hunk 2]
```diff
@@ -69,7 +76,15 @@ public Uri ClientEndpoint
 
         internal string Version { get; }
 
-        internal AccessKey AccessKey { get; private set; }
+        internal AccessKey AccessKey
+        {
+            get
+            {
+                _accessKey ??= new AadAccessKey(_serviceEndpoint, _tokenCredential, ServerEndpoint);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs:83]: this is not thread safe get, does it matter?

File: src/Microsoft.Azure.SignalR.Common/Endpoints/UriExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+using System;
+using System.Linq;
+
+namespace Microsoft.Azure.SignalR
+{
+    internal static class UriExtensions
+    {
+        public static Uri Append(this Uri uri, params string[] paths)
+        {
+            return new Uri(paths.Aggregate(uri.AbsoluteUri, (current, path) => string.Format("{0}/{1}", current.TrimEnd('/'), path.TrimStart('/'))));
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/UriExtensions.cs:13]: use UriBuilder?

**Review by vicancy — APPROVED** [review]

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs [hunk 4]
```diff
@@ -132,16 +147,18 @@ public ServiceEndpoint(string nameWithEndpointType, Uri endpoint, TokenCredentia
...
         {
             _serviceEndpoint = endpoint ?? throw new ArgumentNullException(nameof(endpoint));
             CheckScheme(endpoint);
-            if (credential is null)
-            {
-                throw new ArgumentNullException(nameof(credential));
-            }
-            AccessKey = new AadAccessKey(endpoint, credential);
+
+            _tokenCredential = credential ?? throw new ArgumentNullException(nameof(credential));
```
Comment (Y-Sindo) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs:160]: Why don't we just create a `AadAccessKey` here?

File: src/Microsoft.Azure.SignalR.Common/Endpoints/AadAccessKey.cs (hunk 1 above)
Comment (terencefan) [reply src/Microsoft.Azure.SignalR.Common/Endpoints/AadAccessKey.cs:66]: Port should be included in `endpoint` or `serverEndpoint`

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs (hunk 2 above)
Comment (terencefan) [reply src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs]: fixed

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs [hunk 5]
```diff
@@ -69,7 +78,24 @@ public Uri ClientEndpoint
...
+        {
+            get
+            {
+                if (_accessKey is null)
+                {
+                    lock (_lock)
+                    {
+                        if (_accessKey is null)
+                        {
+                            _accessKey ??= new AadAccessKey(_serviceEndpoint, _tokenCredential, ServerEndpoint);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs:91]: redundant `??`

**Review by vicancy — APPROVED** [review]

**Pull request merged**